src/compiler/utilities.ts

```diff
@@ -444,7 +444,7 @@ namespace ts {
     }
 
     export function isEffectiveExternalModule(node: SourceFile, compilerOptions: CompilerOptions) {
-        return isExternalModule(node) || compilerOptions.isolatedModules;
+        return isExternalModule(node) || compilerOptions.isolatedModules || !!node.commonJsModuleIndicator;

```

Comment:
this seems like the best compromise here: 
`(!!node.commonJsModuleIndicator && getEmitModuleKind(compilerOptions) !== ModuleKind.CommonJS)`

Reply:
so what u think?

Reply:
It seems reasonable. Though anyone with mixed-mode compilations targeting commonjs where they use `require` in some global files is going to get sad when their global files start getting transpiled as modules. Though why we even support mixed mode compilations in the first place...

Reply:
the other option is to do nothing and leave the current behavior.. which is broken on import helpers + js files (which could be a smaller group affected than by the proposed change)...

Reply:
mmmmmmm I think once limited to commonjs it's likely fine. The only difference other than helpers, which is what this aims to change, for module vs non-module would be that top level declarations are local to the file instead of global. And given that the commonjs module flag is only set in JS files, it's not like there'll be any potential interface or namespace merging going on; so it'd just be weather people expected the `var x` they declared in one file with a `require` to be visible in another file with a `require`.